provider/pkg/provider/yaml/v2/yaml.go

```diff
@@ -130,6 +137,41 @@ func yamlDecode(text string, _ *clients.DynamicClientSet) ([]unstructured.Unstru
 	return resources, nil
 }
 
+func Expand(objs []unstructured.Unstructured) ([]unstructured.Unstructured, error) {
+	result := make([]unstructured.Unstructured, 0, len(objs))
+	for {
+		if len(objs) == 0 {
+			break
+		}
+		var obj unstructured.Unstructured
+		obj, objs = objs[0], objs[1:]
+
+		// Recursively traverse built-in Kubernetes list types into a single set of "naked" resource
+		// definitions that we can register with the Pulumi engine.
+		//
+		// Kubernetes does not instantiate list types like `v1.List`. When the API server receives
+		// a list, it will recursively traverse it and perform the necessary operations on
+		// each "instantiable" resource it finds. For example, `kubectl apply` on a
+		// `v1.ConfigMapList` will cause the API server to traverse the list, and `apply` each
+		// `v1.ConfigMap` it finds.
+		//
+		// Since Kubernetes does not instantiate list types directly, Pulumi also traverses lists
+		// for resource definitions that can be managed by Kubernetes, and registers those with the
+		// engine instead.
+		if obj.IsList() {
+			list, err := obj.ToList()
+			if err != nil {
+				return nil, fmt.Errorf("YAML object is invalid: `%s`: %w", printUnstructured(&obj), err)
+			}
+			objs = append(list.Items, objs...)
```

Comment:
I'm slightly worried about expanding list items. There is a long standing bug in upstream kubernetes/kubernetes (https://github.com/kubernetes/kubernetes/issues/3030) where attempting to get the GVK of list items return empty. In cli-utils `HydrateSetList`, it calls `UnstructuredToObjMetadata` to get the object's GVK (https://github.com/kubernetes-sigs/cli-utils/blob/7928dbf3d402196a58ae57d8377961c6dffc8e01/pkg/object/unstructured.go#L32-L38).

I'm not sure if List items would work as expected in the wild - I see you have a unit test added but I'm not sure if it will sufficiently cover this, especially given that this bug occurs for some List objects, not all. We might need a workaround for List objects.

Reply:
After further discussions, this probably isn't a concern since these list objects are generated and consumed locally by Pulumi. It's a concern when we fetch the list objects with client-go from an API server.

Reply:
It is true that this implementation doesn't infer the `kind` of items within a typed list, it expects each item to have a `kind`.  That's true of "yaml/v1" also.

Note that the code uses `unstructured.IsList()` and thus relies on this definition of a list (from [api-conventions.md](https://github.com/chrisnegus/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#types-kinds)):

> Any kind that has the "items" field must be a list kind.